src/lib.rs

```diff
@@ -22,8 +26,19 @@ use regex::Regex;
 use std::fs::{canonicalize, read_to_string};
 use std::path::{Path, PathBuf};
 
-fn resolve_path(path: &str) -> PathBuf {
-    canonicalize(&path).unwrap_or_else(|e| {
+fn resolve_path(path: &str, relative: Option<&Path>) -> PathBuf {
+    if let Some(file) = relative {
+        let p = Path::new(path);
+        if p.has_root() {
+            canonicalize(p.iter().skip(1).collect::<PathBuf>())
+        }
+        else {
+            canonicalize(file.join(p))
+        }
+    }
+    else {
+        canonicalize(path)
+    }.unwrap_or_else(|e| {
```

Comment:
Could you please explain why are we doing `skip(1)` on the path?

Please correct me if I am wrong, but I believe this could be simplified to this:

```suggestion
fn resolve_path(path: &str, parent_dir_path: Option<PathBuf>) -> PathBuf {
    let mut path = PathBuf::from(path);

    if let Some(p) = parent_dir_path {
        if !path.is_absolute() {
            path = p.join(path);
        }
    }

    canonicalize(&path).unwrap_or_else(|e| {
        panic!(
            "An error occured while trying to resolve path: {:?}. Error: {}",
            path, e
        )
    })
}
```

Reply:
There are two cases here:
- The path is relative, and searching should occur based upon the current document.
- The path is "absolute" - which should mean that it is absolute to the crate root. So a path like `/src/temp.glsl` should globally refer to `{CRATE_ROOT}/src/temp.glsl`, no matter where the current file is.

Unfortunately, `path.is_absolute()` does not actually work for the second case. `path.is_absolute` only returns true on Windows for paths that begin with a drive letter or `\\?\`. So the current changes are the (albeit ugly) workaround that I came up with. `p.has_root()` _does_ return true for any path that starts with `/`, so we use that to check whether the path should be crate-root relative, and if it is, then we use `skip(1)` to strip the leading `/`. If `skip(1)` is not used, then `/src/temp.glsl` is interpreted on Windows as `C:\src\temp.glsl`.

I'm happy to refactor the `skip(1)` into a separate function for increased clarity. But I believe that such a workaround is necessary in this case.

Reply:
Thank you for the detailed explanation.

Now wondering if we are trying to handle a case that is not required. Please correct me if I am wrong, but from some quick testing on my Windows machine, it seems like:

- `std::include_str` does not support the leading `/`, it throws an error instead. A relative path is required

- `std::fs::canonicalize` does not support the leading `/`, it throws an error instead. Only relative paths starting with a leading dot slash `./` or a path with no leading `/` are supported, and are resolved from the workspace root directory

Furthermore, the library current version `0.1.1` is using `std::fs::canonicalize` directly to resolve the paths, which would mean that the leading `/` is currently not supported.

Rust source code seems to use `is_absolute` for `std::include_str` to resolve the paths: https://doc.rust-lang.org/nightly/nightly-rustc/src/rustc_expand/base.rs.html#1184-1214

Your thoughts? Any comments or suggestions on this would be greatly appreciated.

Reply:
Interesting - thanks for looking through the standard library implementation. If `include_str` and the like require a relative path, then I believe that we should do the same. I'll go ahead and refactor my code to mirror the linked Rust source :)